source/server/converters.h

```diff
@@ -317,6 +318,19 @@ typename TypeToStorageType<TDriverType, TGrpcType>::StorageType allocate_output_
 }
 
 }  // namespace converters
+
+const int kMaxGrpcErrorDescriptionSize = 4096;
+
+inline ::grpc::Status ApiErrorAndDescriptionToStatus(google::protobuf::int32 status, std::string& description)
+{
+    converters::trim_trailing_nulls(description);
+    // TODO: May need full-fledged json escape logic.
+    for (size_t index = 0; (index = description.find('"', index)) != std::string::npos; ++index) {
+        description.replace(index, 1, "\\\"");
+    }
+    std::string errorMessage = "{ \"code\":\"" + std::to_string(status) + "\", \"" + description + "\" }";
```

Comment:
Google documents that this is a human-readable message:
https://cloud.google.com/apis/design/errors#error_model

There is a separate `repeated Any details` for additional details, but I don't know if we have access to that in all programming languages.

Reply:
status.proto: https://github.com/grpc/grpc/blob/master/src/proto/grpc/status/status.proto
C# RpcException provides a `Status` property: https://github.com/grpc/grpc-dotnet/blob/master/src/Grpc.Core.Api/RpcException.cs#L77
Python provides an `rpc_status.from_call` function: https://github.com/grpc/grpc/blob/master/examples/python/errors/client.py#L36

Reply:
I am aware of that, and while I saw that on their cloud.google.com site, I didn't see that documentation anywhere else. There are numerous problems with it, however:
- The C++ gRPC::Status object `error_details` says it's another serialized gRPC::Status object, such as an inner exception, with the same restrictions.
- The `error_details` don't appear to be propagated to the client exception

Since we're using UNKNOWN, I think we're staying as true to the spirit of the documentation as possible. That field does contain a human-readable message.

Reply:
It's confusing because of the naming, but the Status message listed in that link from Google describing a richer error model than what gRPC provides natively is different from this Status object. What is described in that link is a convention based approach to providing richer error reporting than what you get by default with just an error string and one of the 17 predefined error codes. In particular, it defines several message types which can be included as part of the error to provide more detailed information.

While I had difficulty finding documentation about how this is accomplished, I finally pieced together that this information is passed as part of the metadata in the response trailers using the string "grpc-status-details-bin" as the metadata key. Since this is convention based and not part of gRPC proper, it isn't supported by all languages:

https://www.grpc.io/docs/guides/error/

> This richer error model is already supported in the C++, Go, Java, Python, and Ruby libraries, and at least the grpc-web and Node.js libraries have open issues requesting it. 

However, I have no idea what those libraries look like in each of these languages. There is also an open issue to support this for C#. In the meantime, I did find these examples of how it can be done from C#:

https://github.com/chwarr/grpc-dotnet-google-rpc-status
https://github.com/nano-byte/grpc-rich-error

Whether it would be better for NI to adopt this convention in some NI standard way is something I've been mulling over for Measurement SDK. Not all of the message classes defined by Google work well for us, but we could easily publish our own message types that do. Also, C# and LabVIEW don't currently have libraries available that make accessing the detailed error information easy. LabVIEW currently doesn't even support gRPC metadata at all.

For now, maybe the best thing is to concede to encoding driver error message/code into a json string that is returned as the error message reported natively by gRPC. That's probably human readable enough to be useful for most users. We can always decide to return more detailed error information later using the convention described above which may duplicate some of the error information included in the standard error message.

Reply:
Thanks for the correction @jasonmreding. I mixed up the hand-written `Status` and the code-generated `Status`.

In Python, on the client, you call `rpc_status.from_call(call)` to unpack the call's `grpc-status-details-bin` response trailer into a `status_pb2.Status` (the codegen from `status.proto`). It also validates that the error code and message match.
https://github.com/grpc/grpc/blob/master/src/python/grpcio_status/grpc_status/rpc_status.py#L33

> For now, maybe the best thing is to concede to encoding driver error message/code into a json string that is returned as the error message reported natively by gRPC. That's probably human readable enough to be useful for most users.

JSON allows adding fields without breaking compatibility, but switching from JSON to "richer error model" would be a breaking change.

Other ways to represent the error code without JSON:
- "Error -50150: The software has entered an unknown state..." (English-specific)
- "-50150: The software has entered an unknown state..." (language-agnostic, but more cryptic)

Reply:
> JSON allows adding fields without breaking compatibility, but switching from JSON to "richer error model" would be a breaking change.

When I said switch I meant also use the richer error model. When using the richer error model, you still need to report error information through the standard gRPC mechanism. Presumably we would continue to report errors through the standard mechanism as is done in this PR (not break it). I'm not actually sure how the standard mechanism encodes that data over http. I'm guessing it's either part of the response body or trailing headers. If it's headers, I'm not sure what the metadata key is that is used.

> Other ways to represent the error code without JSON:

My concern with that format, is it makes it very difficult to trap/handle specific errors programmatically. With JSON, you can at least get at the error code in a standard way even though it's more work.

Reply:
Looks like standard error info is sent as trailing headers using "grpc-status" and "grpc-message".

https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses

Reply:
Note that the Python example you linked to is experimental, which means it's subject to change. And the only example I remember seeing was getting the metadata by expanding both parameters at the call site, which meant it wouldn't throw like we want it to. Or at least that's what it looked like to me.

Jason also the C# examples that do have richer fields are only defined for certain error types. There's no obvious way to extend it for others such as UNKNOWN.

Reply:
> Jason also the C# examples that do have richer fields are only defined for certain error types. There's no obvious way to extend it for others such as UNKNOWN.

I'm not sure what examples you are referring to. I'm not aware of any conventions that say this extended error model should only be used with certain grpc status codes. To my knowledge, it would still be up to us to decide on a convention for which grpc status code to return in a given situation. My thought is that most things would get mapped to UNKNOWN. I'm not sure if there is value in sometimes reporting something like INVALID_ARGUMENT vs. just always reporting UNKNOWN. The client receiving the error ultimately has to do something:

``` CSHARP
if (errorDetails is MessageType1) { /* display custom error info for message1 */}
else if (errorDetails is MessageType2) {/* display custom error info for message2 */}
else {/* generic error display fallback */}
```
so maybe the status codes could help minimize the number of checks for different types of error info? I think the more likely answer is we'll have to do better about documenting which types of extended error information can be returned from a given rpc, basically the equivalent of documenting which types of exceptions can be thrown from a function.

Reply:
> When I said switch I meant also use the richer error model. When using the richer error model, you still need to report error information through the standard gRPC mechanism. Presumably we would continue to report errors through the standard mechanism as is done in this PR (not break it).

Okay, that makes sense.

> I'm not sure what examples you are referring to

I linked to a Python example that uses an experimental API.

> I'm not aware of any conventions that say this extended error model should only be used with certain grpc status codes. 

I'm not aware of any either. 

One of the error detail messages that Google provides is `DebugInfo`, which adds a server-side stack trace. This seems applicable to all errors.

> And the only example I remember seeing was getting the metadata by expanding both parameters at the call site, which meant it wouldn't throw like we want it to. Or at least that's what it looked like to me.

gRPC attaches the response trailer metadata to the exception. I'm not sure how this works in Python, but it's pretty clear in C#.

There is a 3rd party C# library implementing the richer error model: https://github.com/nano-byte/grpc-rich-error

> The client receiving the error ultimately has to do something:
> 
> ```
> if (errorDetails is MessageType1) { /* display custom error info for message1 */}
> else if (errorDetails is MessageType2) {/* display custom error info for message2 */}
> else {/* generic error display fallback */}
> ```

`details` is a `repeated Any`, so you can include multiple detail objects.

To extend the example from nano-byte's C# library:
```
try
{
    // ...
}
catch (RpcException ex)
{
    var info = ex.GetDetail<ErrorInfo>();
    if (info != null)
    {
        // ...
    }
    var debugInfo = ex.GetDetail<DebugInfo>();
    if (debugInfo != null)
    {
        // print stack trace...
    }
}
```